src/Rules.hs

```diff
@@ -20,33 +21,42 @@ import qualified Rules.Program
 import qualified Rules.Register
 import Settings
 import Target
-import UserSettings
 import Utilities
 
 allStages :: [Stage]
-allStages = [minBound ..]
+allStages = [minBound .. maxBound]
 
 -- | This rule calls 'need' on all top-level build targets, respecting the
 -- 'Stage1Only' flag.
 topLevelTargets :: Rules ()
-topLevelTargets = action $ do
-    let libraryPackages = filter isLibrary (knownPackages \\ [libffi])
-    need =<< if stage1Only
-             then do
-                 libs <- concatForM [Stage0, Stage1] $ \stage ->
-                     concatForM libraryPackages $ packageTargets False stage
-                 prgs <- concatForM programsStage1Only $ packageTargets False Stage0
-                 return $ libs ++ prgs ++ inplaceLibCopyTargets
-             else do
-                 targets <- concatForM allStages $ \stage ->
-                     concatForM (knownPackages \\ [libffi]) $
-                        packageTargets False stage
-                 return $ targets ++ inplaceLibCopyTargets
+topLevelTargets = do
+    phony "stage2" $ do
+      putNormal "Building stage2"
+      (programs, libraries) <- partition isProgram <$> stagePackages Stage1
+      pgmNames <- mapM (g Stage1) programs
+      libNames <- mapM (g Stage1) libraries
+      putNormal . unlines $
```

Comment:
Should this be under some `verbosity` conditional?

Reply:
Presumably the output is quite long and almost never changes, so the user will unlikely want to always see this information.

Reply:
Good idea. Done in an upcoming commit.